include/zio/zio_dev.h

```diff
@@ -129,75 +177,76 @@ struct zio_dev_chan {
+		ZIO_BYTEORDER_LITTLE
 	} byte_order;
 
 	const enum sign_bit {
 		ZIO_SIGN_NONE,
-			ZIO_SIGN_MSB,
-			ZIO_SIGN_LSB,
+		ZIO_SIGN_MSB,
+		ZIO_SIGN_LSB,
 	} sign_bit;
-
 };
 
+
 /**
- * @brief Function to set a device attribute
+ * @brief Function to set a channel attribute for a device
  *
- * @return -EINVAL if an invalid attribute id was given, 0 otherwise
+ * @return -EINVAL if an invalid attribute id, type, or range was given,
+ * 0 otherwise
  */
-typedef int (*zio_dev_set_attr_t)(struct device *dev, const u32_t attr_idx,
-				  const struct zio_variant var);
+typedef int (*zio_set_attr_t)(struct device *dev, const u16_t channel_idx, const u16_t attribute_idx, const struct zio_variant var);
 
 /**
- * @brief Function to get a device attribute
+ * @brief Function to get a device attribute channel
  *
  * @return -EINVAL if an invalid attribute id was given, 0 otherwise
  */
-typedef int (*zio_dev_get_attr_t)(struct device *dev, uint32_t attr_idx,
-				  struct zio_variant *attr);
+typedef int (*zio_get_attr_t)(struct device *dev, const u16_t channel_idx, const u16_t attribute_idx, struct zio_variant *attr);
+
+
+/** Attribute record. */
+struct zio_attr_desc {
+	/* Attribute data type describing the expected data type. */
+	enum zio_variant_type data_type;
+	zio_set_attr_t set_attr;
+	zio_get_attr_t get_attr;
+};
 
 
 /**
  * @brief Function to get a pointer to the array of attribute descriptions
  */
-typedef int (*zio_dev_get_attrs_t)(struct device *dev,
-				   struct zio_attr **attrs,
-				   u16_t *num_attrs);
+typedef int (*zio_dev_get_attr_descs_t)(struct device *dev,
+					const struct zio_attr_desc **attrs,
+					u32_t *num_attrs);
 
 /**
  * @brief Function to get a pointer to the array of channel descriptions
  */
-typedef int (*zio_dev_get_chans_t)(struct device *dev,
-				   const struct zio_dev_chan **chans,
-				   u32_t *num_chans);
+typedef int (*zio_dev_get_chan_descs_t)(struct device *dev,
+					const struct zio_chan_desc **chans,
+					u32_t *num_chans);
 
 /**
  * @brief Function to get a channels attribute descriptions
  *
  * @return -EINVAL if an invalid channel id was given, 0 otherwise
  */
-typedef int (*zio_chan_get_attrs_t)(struct device *dev, const u32_t chan_idx,
-				    struct zio_attr *attrs,
-				    u32_t *num_attrs);
+typedef int (*zio_chan_get_attr_descs_t)(struct device *dev,
+					 const u32_t chan_idx,
+					 const struct zio_attr_desc **attrs,
+					 u32_t *num_attrs);
+
 
-/**
- * @brief Function to set a channel attribute for a device
- *
- * @return -EINVAL if an invalid attribute id, type, or range was given,
- * 0 otherwise
- */
-typedef int (*zio_chan_set_attr_t)(struct device *dev, const u32_t chan_idx,
-				   const u32_t attr_idx,
-				   const struct zio_variant var);
 /**
  * @brief Function to get a channel attribute for a device
  *
  * @return -EINVAL if an invalid attribute id was given, 0 otherwise
  */
 typedef int (*zio_chan_get_attr_t)(struct device *dev, const u32_t chan_idx,
```

Comment:
I don't think this, or any of the other get/set attr function pointer defs are needed anymore with your change, where each attribute has its own getter/setter